### PR TITLE
Add floor and wall openings both working, issue templates added

### LINKED ISSUE #77: RAM_Toolkit: Create functionality for wall openings and floor openings
Needs some RAM gymnastics to get openings to come through for panels, depending on type. May come in as construction lines for floors due to unsupported API methods.

### LINKED ISSUE #108: RAM_Toolkit: Issue templates
**By 30th November please**

Please include issue and pull request templates for this repository :smile:

### PR DESCRIPTION
**Pull request by enarhi**
### Issues addressed by this PR

Closes #77 
Closes #108 

Add wall openings now supported, add slab and opening also supported.

### Changelog

- Added functionality for creating wall openings
- Improved slab creation and added slab opening creation

#### Additional notes
Test script: https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/RAM_Toolkit/Issue77%20Floor%20and%20Wall%20Openings?csf=1&e=CkcfEq

### CODE REVIEW

File: RAM_Adapter/CRUD/Create.cs [hunk 1]
```diff
@@ -453,18 +483,41 @@ private bool CreateCollection(IEnumerable<Panel> bhomPanels)
...
                         {
                             ramFloorType = ramStory.GetFloorType();
-                            //Get ILayoutWalls of FloorType
+                            
+                            //Get ILayoutWalls of FloorType and add wall
                             ILayoutWalls ramLayoutWalls = ramFloorType.GetLayoutWalls();
+                            ILayoutWall ramLayoutWall = ramLayoutWalls.Add(EMATERIALTYPES.EWallPropConcreteMat, wallMin.X, wallMin.Y, 0, 0, wallMax.X, wallMax.Y, 0, 0, thickness);
+
+                            //Set lateral
+                            ramLayoutWall.eFramingType = EFRAMETYPE.MemberIsLateral;
```
[the remark this answers is not in the document]
Comment (enarhi) [reply RAM_Adapter/CRUD/Create.cs]: T follow in a later issue.

File: RAM_Adapter/CRUD/Create.cs [hunk 2]
```diff
@@ -258,29 +258,53 @@ private bool CreateCollection(IEnumerable<IMaterialFragment> materials)
...
+                    {
+                        ICompDeckProp ramProp = ramCompDeckProps.GetAt(i);
+                        if (ramProp.strLabel == srfProp.Name)
+                        {
+                            existingProp = ramProp.lUID;
+                        }
+                    }
 
-            //    ICompDeckProps.Add(deckName, thickness, studLength);
+                    if (existingProp == 0)
```
[the remark this answers is not in the document]
Comment (enarhi) [reply RAM_Adapter/CRUD/Create.cs]: Agree, to follow in a compliance issue with the ID updates.

File: RAM_Adapter/CRUD/UpdateObjects.cs [hunk 3]
```diff
@@ -74,7 +74,7 @@ protected bool Update(IEnumerable<Bar> bars)
             foreach (Bar bar in bars)
             {
                 //Extract ID
-                int ID = System.Convert.ToInt32(bar.CustomData["lUID"]);
+                int ID = System.Convert.ToInt32(bar.CustomData[AdapterId]);
```
[the remark this answers is not in the document]
Comment (enarhi) [reply RAM_Adapter/CRUD/UpdateObjects.cs]: To follow in a compliance issue with the ID updates.

**Comment by enarhi** [comment]
New test file added to: https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/RAM_Toolkit/Issue77%20Floor%20and%20Wall%20Openings?csf=1&e=CUepQs

191208...

Please check with that one, I am getting all openings coming through on my end as below:
image

image
image

**Comment by enarhi** [comment]
For overlapping edges on openings on the floors the openings are eliminated as below:
image

However, the opening comes through, its just a matter of how RAM handles it. Suggest this be moved to a separate issue as it is a unique case that should be avoided anyhow. Looks like the rest of the openings create options are working, if you can confirm on your end @JosefTaylor then we can go ahead and get this merged up.

**Comment by JosefTaylor** [comment]
Looks good to me. I also tried multiple random openings. Maybe as a separate issue we can get a good error message about openings which are collinear with external edges.

**Pull request merged**